### PR TITLE
test: generate native function type hint checks

### LINKED ISSUE #159259: Use aten/src/ATen/native/native_functions.yaml to automatically generate tests for type hints
## Summary

Use the information in aten/src/ATen/native/native_functions.yaml to automatically create and maintain "op type tests": Python files proving to `mypy` or `pyrefly` that pytorch's ops have type annotations compatible with the actual type descriptions from `native_functions.yaml`.

PyTorch already has some "type tests" [here](https://github.com/pytorch/pytorch/tree/main/test/typing/pass) but they are small and not and not provably up-to-date with `native_functions.yaml`.

## An example for a single op

Let's randomly choose [`torch.randint`](https://docs.pytorch.org/docs/stable/generated/torch.randint.html) as a "typical op" that has [several variants](https://github.com/pytorch/pytorch/blob/8e07c9870d07c5a318ab21bb16b3fa27576851e6/aten/src/ATen/native/native_functions.yaml#L4744-L4783).

For each of these eight variants, there are a rather large number of Python unit test cases automatically created and run, tests which cover almost all permutations of the possible parameters to this `torch.randint` variant

This feature  proposes "op type tests" for `randint` and all other ops and variants: a Python file, automatically created and updated, intended for `mypy` or `pyrefly` only, to make sure that the type signature that we provide to type checkers is compatible with the variants as specified in `native_functions.yaml`.

Zeroing in on `torch.randint`'s [primary variant](https://github.com/pytorch/pytorch/blob/8e07c9870d07c5a318ab21bb16b3fa27576851e6/aten/src/ATen/native/native_functions.yaml#L4744-L4747), this would generate "type test code" looking like this:

```
class Broken:
    pass
BROKEN = Broken()

INT = 4
SIZE = [1, 2]
DTYPE = torch.int8
LAYOUT = ....

torch.randint(INT, SIZE)
torch.randint(INT, SIZE, dtype=DTYPE, layout=LAYOUT, device=DEVICE, pin_memory=BOOL)
# 2**4 - 2 other combinations are possible, but would add little

torch.randint(BROKEN, SIZE)  # type: ignore[arg-type]
torch.randint(INT, BROKEN)  # type: ignore[arg-type]
torch.randint(INT, SIZE, dtype=BROKEN)  # type: ignore[arg-type]
torch.randint(INT, SIZE, layout=BROKEN)  # type: ignore[arg-type]
torch.randint(INT, SIZE, device=BROKEN)  # type: ignore[arg-type]
torch.randint(INT, SIZE, pin_memory=BROKEN)  # type: ignore[arg-type]
# 2**6 - 6 other combinations are possible, but would add little
```

## How to organize all these type tests

For a single file type test, there are 2653 ops including variants at this moment, so if each of these generates a dozen lines like this, that's about 32kloc in that one file.

But 2653 files at one per file is a lot, or even just 1538 if we put variants together.

Organizing in files by first alphabetic character in the opname would give this many ops per file.

`[('a', 106), ('b', 43), ('c', 140), ('d', 36), ('e', 47), ('f', 187), ('g', 35), ('h', 32), ('i', 64), ('j', 1), ('k', 4), ('l', 118), ('m', 97), ('n', 85), ('o', 8), ('p', 36), ('q', 19), ('r', 80), ('s', 223), ('t', 83), ('u', 51), ('v', 26), ('w', 11), ('x', 3), ('z', 3)]`

so the longest file would be ~2.5 kloc.

Suggestion: go with the really large, single file; if this ends up being some sort of bottleneck, split by first character of opname.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wgu9**
## Issue

Fixes #159259

## Summary

Adds a small generator for native-function type hint reveal tests. The generator reads `native_functions.yaml` through the same torchgen/pyi signature path used by `tools/pyi/gen_pyi.py`, derives expected return reveals from generated signatures, and emits a representative `test/typing/reveal/native_functions.py` file.

The initial generated coverage is intentionally scoped to 13 Tensor-returning cases across unary ops, factories, shape/list arguments, out arguments, dtype/device kwargs, dim tuples, indexing, view-style ops, and matrix ops. A freshness test keeps the checked-in reveal file synchronized with the generator.

## Testing

- `python tools/typing/gen_native_function_type_tests.py`
- `python -m ruff check tools/typing/gen_native_function_type_tests.py test/test_typing.py test/typing/reveal/native_functions.py`
- `python -m py_compile tools/typing/gen_native_function_type_tests.py test/test_typing.py`
- `git diff --check`
- generator freshness smoke check: rendered output equals `test/typing/reveal/native_functions.py` with 13 reveal cases

I could not run the full `test/test_typing.py` mypy suite locally because this checkout is not built as a source tree and the local Python environment is missing `mypy`/`expecttest`.

## Checklist

- [ ] Passes lint (`spin fixlint`)
- [x] Added/updated tests
- [ ] Updated documentation (if applicable)
- [ ] Included benchmark results (for PRs impacting perf)

## BC-breaking?

No.

### CODE REVIEW

**Comment by wgu9** [comment]
EasyCLA should now be signed for this account; requesting a recheck.

File: tools/typing/gen_native_function_type_tests.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    NativeFunctionTypeTestCase("reshape", "torch.reshape(x, (3, 2))"),
+    NativeFunctionTypeTestCase("permute", "torch.permute(cube, (2, 0, 1))"),
+    NativeFunctionTypeTestCase("transpose", "torch.transpose(matrix, 0, 1)"),
+    NativeFunctionTypeTestCase(
+        "sum", "torch.sum(matrix, dim=1, keepdim=True, dtype=torch.float64)"
+    ),
+    NativeFunctionTypeTestCase("amax", "torch.amax(cube, dim=(1, 2), keepdim=True)"),
+    NativeFunctionTypeTestCase(
+        "empty",
+        "torch.empty((2, 3), dtype=torch.float32, device=torch.device('cpu'))",
```
Comment (rec) [inline tools/typing/gen_native_function_type_tests.py]: Tiny question: why is just this one marked as `cpu`?

Comment (wgu9) [reply tools/typing/gen_native_function_type_tests.py]: Good point. It only needed to exercise the device keyword path, not a concrete CPU device. I changed the generated case to use device=None so the reveal test stays backend-neutral while still covering the optional device kwarg.

File: tools/typing/gen_native_function_type_tests.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    deprecated_yaml_path: Path = DEFAULT_DEPRECATED_YAML,
+    test_cases: Sequence[NativeFunctionTypeTestCase] = TEST_CASES,
+) -> str:
+    return_types = get_native_function_return_types(
+        native_yaml_path=native_yaml_path,
+        tags_yaml_path=tags_yaml_path,
+        deprecated_yaml_path=deprecated_yaml_path,
+    )
+
+    lines = [
```
Comment (rec) [inline tools/typing/gen_native_function_type_tests.py]: On second look, one more comment:

Likely these this constant `lines` will be created once and maintained one hundred times, so let's make it easy on those maintainers and make this a multiline string constant at the top-level, and use splitlines:

```
NATIVE_FUNCTIONS_HEADER = """\
# mypy: disable-error-code="possibly-undefined"
# flake8: noqa
# This file is generated by tools/typing/gen_native_function_type_tests.py.
# Do not edit manually; update TEST_CASES in that generator.
import torch


x = torch.randn(2, 3)
matrix = torch.randn(2, 3)
matrix_t = torch.randn(3, 2)
cube = torch.randn(2, 3, 4)
indices = torch.tensor([0, 1], dtype=torch.long)

"""
```

Comment (wgu9) [reply tools/typing/gen_native_function_type_tests.py]: Agreed, that is easier to maintain. I moved the fixed generated-file prelude into a top-level NATIVE_FUNCTIONS_HEADER multiline string and now initialize the generated lines with splitlines(). I also reran the generator and pushed the synchronized reveal file.

**Review by rec — APPROVED** [review]
Sorry about the delay, looks good to me.

I'm surprised you remembered my discussions of this idea, it was quite a while ago. Thanks for keeping me in the loop.

**Comment by wgu9** [comment]
This PR is still relevant. The linked issue is open, and current main still does not contain the native function typing generator or generated reveal test. It is approved and mergeable, but I do not have permission to remove the `Stale` label. Could a maintainer remove it when convenient?

**Comment by rec** [comment]
Don't stress: if it does get closed, you can just re-open it with no prejudice, unless things have changed since I worked on this project.
